optuna/pruners/percentile.py

```diff
@@ -93,10 +99,18 @@ def prune(self, study, trial):
         if step is None:
             return False
 
-        if step <= self.n_warmup_steps:
+        n_warmup_steps = self.n_warmup_steps
+        if step <= n_warmup_steps:
             return False
 
-        if len(trial.intermediate_values) == 0:
+        lower_bound = (
+            (step - n_warmup_steps - 1) // self.interval_steps * self.interval_steps
+            + n_warmup_steps + 1)
+        assert lower_bound >= 0
+        second_last_step = six.moves.reduce(
+            lambda second_last_step, s: s if s > second_last_step and s != step
+            else second_last_step, trial.intermediate_values.keys(), -1)
+        if lower_bound <= second_last_step:
             return False
```

Comment:
Please extract the logic about pruning intervals to a private method or function to improve the code readability. I feel the `prune` method is getting complicated and it takes a few minutes to find the logic.

Reply:
I agree. Refactored it out to a function `_is_first_step_after_nearest_lower_pruning_step`, with a somewhat self-documented name.